tests/binary_agreement.rs

```diff
@@ -27,31 +27,139 @@ extern crate threshold_crypto as crypto;
+    known_node_ids: Vec<NodeId>,
+    /// Node ids under control of adversary.
+    known_adversarial_ids: Vec<NodeId>,
+    /// The most recent epoch we've seen.
+    current_epoch: u32,
+    /// If we've sent the one-time messages for the attack.
+    sent_attack: bool,
+}
+
+impl Adversary<BinaryAgreement<NodeId>> for ReorderingAdversary {
+    fn init(
+        &mut self,
+        all_nodes: &BTreeMap<NodeId, TestNode<BinaryAgreement<NodeId>>>,
+        nodes: &BTreeMap<NodeId, Arc<NetworkInfo<NodeId>>>,
+    ) {
+        self.known_adversarial_ids = nodes.keys().cloned().collect();
+        self.known_node_ids = all_nodes.keys().cloned().collect();
+    }
+
+    fn pick_node(&self, nodes: &BTreeMap<NodeId, TestNode<BinaryAgreement<NodeId>>>) -> NodeId {
+        // Prioritizes A0, as required by the attack.
+        MessageScheduler::First.pick_node(nodes)
+    }
+
+    fn push_message(&mut self, _: NodeId, msg: TargetedMessage<Message, NodeId>) {
+        if self.current_epoch < msg.message.epoch {
+            self.current_epoch = msg.message.epoch;
+            // Reset for the next epoch.
+            self.sent_attack = false;
+        }
+    }
+
+    fn step(&mut self) -> Vec<MessageWithSender<BinaryAgreement<NodeId>>> {
+        let mut messages = Vec::new();
+        if !self.sent_attack {
+            self.sent_attack = true;
+            for &our_node_id in &self.known_adversarial_ids {
+                let f = self.known_node_ids.len() / 3;
+                for &target in &self.known_node_ids[0..f] {
+                    // Disagree with the first third.
+                    messages.push(MessageWithSender::new(
+                        our_node_id,
+                        TargetedMessage {
+                            target: Target::Node(target),
+                            message: Message {
+                                epoch: self.current_epoch,
+                                content: MessageContent::SbvBroadcast(
+                                    sbv_broadcast::Message::BVal(true),
+                                ),
+                            },
+                        },
+                    ));
+                }
+                for &target in &self.known_node_ids[f..(f * 2)] {
+                    // Affirm the second third.
+                    messages.push(MessageWithSender::new(
+                        our_node_id,
+                        TargetedMessage {
+                            target: Target::Node(target),
+                            message: Message {
+                                epoch: self.current_epoch,
+                                content: MessageContent::SbvBroadcast(
+                                    sbv_broadcast::Message::BVal(false),
+                                ),
+                            },
+                        },
+                    ));
+                }
+            }
+        }
+        messages
```

Comment:
This doesn't really reorder any messages, it just forges new messages from correct nodes, which the adversary can't do in reality. These messages may then arrive in any order (_in addition_ to the original ones, sent by the correct nodes themselves), so I don't think it reliably replicates the attack.

I think the new `net` framework provides better controls over message reordering. (Please correct me if I'm wrong, @mbr.)

Also, the attacker needs to reconstruct the coin value before anyone else, and then send the `Aux` messages accordingly.

Once the test is complete, please try (without checking it in, or maybe just in a separate temporary branch) disabling the `Conf` messages in `binary_agreement`, and use `aux_vals` itself instead of `conf_values` to try and update the epoch. With that change, this test should fail (i.e. the attack works) but all the other tests should still pass (i.e. `binary_agreement` is still working as originally described in the paper).